### PR TITLE
lib/strings: add parseLinuxConfig, linuxManualConfig: refactor readConfig

### PR DESCRIPTION
**Pull request by linyinfeng**
## Description of changes

With this PR, when `bulitins.readFile configfile` is not IFD (import from derivation), kernel packages using `linuxManualConfig` can be evaluated without IFD. For example, when `configfile = ./config`.

This PR:
1. Add a new function `parseLinuxConfig` to `lib.strings`.
2. Refactor `readConfig` in `linuxManualConfig` with the newly added function.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by shlevy** [comment]
Would be curious to see some profiling results of this vs the IFD approach on a real config.

**Comment by lolbinarycat** [comment]
one advantage of IFD is it actually allows expensive computations to be cached.  not sure if parsing a trivial format counts as "expensive computation", but it's something to keep in mind when benchmarking.

**Comment by linyinfeng** [comment]
> Would be curious to see some profiling results of this vs the IFD approach on a real config.

> one advantage of IFD is it actually allows expensive computations to be cached. not sure if parsing a trivial format counts as "expensive computation", but it's something to keep in mind when benchmarking.

I have done some benchmarks with `hyperfine`. Performance with `parseLinuxConfig` is almost the same as the performance of the IFD approach when `config.nix` has been cached (on my laptop with the kernel config of my laptop).

```
Benchmark 1: nix eval .#lib.configOriginal --option eval-cache false >/dev/null
  Time (mean ± σ):     330.8 ms ±  38.7 ms    [User: 218.0 ms, System: 61.3 ms]
  Range (min … max):   303.1 ms … 435.4 ms    10 runs
 
Benchmark 2: nix eval .#lib.configNew --option eval-cache false >/dev/null
  Time (mean ± σ):     314.2 ms ±   8.1 ms    [User: 230.9 ms, System: 57.0 ms]
  Range (min … max):   304.7 ms … 328.6 ms    10 runs
```

Here is the benchmark script:
 <https://gist.github.com/linyinfeng/099b6061a08fc381e28c7a3eaa612f01>.

```nix
{
  inputs = { nixpkgs-299926.url = "github:nixos/nixpkgs/pull/299926/head"; };
  outputs = { self, nixpkgs-299926 }:
    let
      testSystem = "x86_64-linux";
      pkgs = nixpkgs-299926.legacyPackages.${testSystem};
      inherit (pkgs) runCommand lib writeShellApplication;
      inherit (builtins) toJSON;
      # from https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kernel/manual-config.nix
      readConfigOriginal = configfile:
        import (runCommand "config.nix" { } ''
          echo "{" > "$out"
          while IFS='=' read key val; do
            [ "x''${key#CONFIG_}" != "x$key" ] || continue
            no_firstquote="''${val#\"}";
            echo '  "'"$key"'" = "'"''${no_firstquote%\"}"'";' >> "$out"
          done < "${configfile}"
          echo "}" >> $out
        '').outPath;
      # from https://github.com/NixOS/nixpkgs/pull/299926/files
      readConfigNew = configfile:
        lib.strings.parseLinuxConfig {
          # IFD could happen in `builtins.readFile`
          configText = builtins.readFile configfile;
          relaxed = true;
        };
    in {
      lib = {
        configOriginal = readConfigOriginal ./example.config;
        configNew = readConfigNew ./example.config;
      };
      packages.${testSystem} = {
        config-diff = runCommand "config-diff" { } ''
          diff <(${toJSON self.lib.configOriginal}) <(${
            toJSON self.lib.configNew
          }) > $out
        '';
        test = writeShellApplication {
          name = "test";
          runtimeInputs = with pkgs; [ hyperfine ];
          text = let
            # disable flake eval cache
            cmd = item:
              "nix eval .#lib.${item} --option eval-cache false >/dev/null";
            cmdEmpty =
              "nix eval --expr '{}' --option eval-cache false >/dev/null";
          in ''
            echo -n "lines: "
            wc -l example.config
            echo "-- original: first time eval --"
            time ${cmd "configOriginal"}
            echo "-- new:      first time eval --"
            time ${cmd "configNew"}
            echo "-- hyperfine (--warmup 3) --"
            hyperfine --warmup 3 "${cmd "configOriginal"}" "${
              cmd "configNew"
            }" "${cmdEmpty}"
          '';
        };
      };
    };
}
```

Note that I passed `--option eval-cache false` to disable flake eval cache.

Here are the full benchmark console logs. I benchmarked twice. In the first benchmark `configNew` is slightly faster than `configOriginal`, and in the second benchmark `configOriginal` is slightly faster.

```console
yinfeng@framework ~/S/p/clean> nix --version
nix (Nix) 2.18.2
yinfeng@framework ~/S/p/clean> curl https://gist.githubusercontent.com/linyinfeng/099b6061a08fc381e28c7a3eaa612f01/raw/8f60ab7dad1403c9066282c6660d391d440c8725/flake.nix > flake.nix
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2313  100  2313    0     0   4427      0 --:--:-- --:--:-- --:--:--  4431
yinfeng@framework ~/S/p/clean> nix flake lock
warning: creating lock file '/home/yinfeng/Source/playground-benchmark-parse-linux-config/clean/flake.lock'
yinfeng@framework ~/S/p/clean> cat /proc/config.gz | gzip -d > example.config
yinfeng@framework ~/S/p/clean> head example.config
#
# Automatically generated file; DO NOT EDIT.
# Linux/x86_64 6.6.14 Kernel Configuration
#
CONFIG_CC_VERSION_TEXT="gcc (GCC) 13.2.0"
CONFIG_CC_IS_GCC=y
CONFIG_GCC_VERSION=130200
CONFIG_CLANG_VERSION=0
CONFIG_AS_IS_GNU=y
CONFIG_AS_VERSION=24100
yinfeng@framework ~/S/p/clean> echo "# some random comment, so that we can test the first time performance" >> example.config 
yinfeng@framework ~/S/p/clean> tail example.config 
CONFIG_MEMTEST=y
# CONFIG_HYPERV_TESTING is not set
# end of Kernel Testing and Coverage

#
# Rust hacking
#
# end of Rust hacking
# end of Kernel hacking
# some random comment, so that we can test the first time performance
yinfeng@framework ~/S/p/clean> wc -l example.config
11712 example.config
yinfeng@framework ~/S/p/clean> nix run .#test
lines: 11712 example.config
-- original: first time eval --

real	0m6.105s
user	0m0.221s
sys	0m0.084s
-- new:      first time eval --

real	0m0.332s
user	0m0.231s
sys	0m0.074s
-- hyperfine (--warmup 3) --
Benchmark 1: nix eval .#lib.configOriginal --option eval-cache false >/dev/null
  Time (mean ± σ):     330.8 ms ±  38.7 ms    [User: 218.0 ms, System: 61.3 ms]
  Range (min … max):   303.1 ms … 435.4 ms    10 runs
 
Benchmark 2: nix eval .#lib.configNew --option eval-cache false >/dev/null
  Time (mean ± σ):     314.2 ms ±   8.1 ms    [User: 230.9 ms, System: 57.0 ms]
  Range (min … max):   304.7 ms … 328.6 ms    10 runs
 
Benchmark 3: nix eval --expr '{}' --option eval-cache false >/dev/null
  Time (mean ± σ):      16.1 ms ±   1.5 ms    [User: 9.7 ms, System: 5.9 ms]
  Range (min … max):    13.1 ms …  20.9 ms    170 runs
 
Summary
  nix eval --expr '{}' --option eval-cache false >/dev/null ran
   19.53 ± 1.94 times faster than nix eval .#lib.configNew --option eval-cache false >/dev/null
   20.57 ± 3.12 times faster than nix eval .#lib.configOriginal --option eval-cache false >/dev/null
yinfeng@framework ~/S/p/clean> nix build .#config-diff
yinfeng@framework ~/S/p/clean> cat ./result
yinfeng@framework ~/S/p/clean> # diff is empty!
yinfeng@framework ~/S/p/clean> nix run .#test
lines: 11712 example.config
-- original: first time eval --

real	0m0.322s
user	0m0.215s
sys	0m0.064s
-- new:      first time eval --

real	0m0.535s
user	0m0.280s
sys	0m0.060s
-- hyperfine (--warmup 3) --
Benchmark 1: nix eval .#lib.configOriginal --option eval-cache false >/dev/null
  Time (mean ± σ):     315.7 ms ±   8.0 ms    [User: 206.3 ms, System: 68.0 ms]
  Range (min … max):   303.7 ms … 326.1 ms    10 runs
 
Benchmark 2: nix eval .#lib.configNew --option eval-cache false >/dev/null
  Time (mean ± σ):     316.4 ms ±  14.6 ms    [User: 230.3 ms, System: 61.0 ms]
  Range (min … max):   304.0 ms … 352.6 ms    10 runs
 
Benchmark 3: nix eval --expr '{}' --option eval-cache false >/dev/null
  Time (mean ± σ):      15.9 ms ±   1.5 ms    [User: 10.0 ms, System: 5.5 ms]
  Range (min … max):    13.1 ms …  20.5 ms    178 runs
 
Summary
  nix eval --expr '{}' --option eval-cache false >/dev/null ran
   19.83 ± 1.99 times faster than nix eval .#lib.configOriginal --option eval-cache false >/dev/null
   19.88 ± 2.14 times faster than nix eval .#lib.configNew --option eval-cache false >/dev/null
```

**Comment by linyinfeng** [comment]
I found that I made some mistakes in `checks.<system>.config-diff`. But the conclusion does not change. I have updated the code in gist.

```nix
      packages.${testSystem} = let
        formattedJSON = name:
          pkgs.runCommand name { } ''
            ${pkgs.jq}/bin/jq <"${
              toFile "original" (toJSON self.lib.${name})
            }" > $out
          '';
      in {
        config-diff = runCommand "config-diff" { } ''
          diff "${formattedJSON "configOriginal"}" "${formattedJSON "configNew"}"
          touch $out
        '';
```

**Comment by infinisil** [comment]
Hmm not a fan of this.

Here's an alternative idea: How about exposing the [tool to convert a kernel config to a `config.nix`](https://github.com/NixOS/nixpkgs/blob/9ce47d84cddede3ee9330f63f458e60e267f939c/pkgs/os-specific/linux/kernel/manual-config.nix#L11-L19) somewhere under `pkgs.*`, such that users can commit a `config.nix` instead. And from a `config.nix` it's very straightforward to generate a kernel config again.

File: lib/strings.nix [hunk 1]
```diff
@@ -1149,6 +1150,68 @@ rec {
...
+    configText,
+    # skip ill-formed lines
+    relaxed ? false,
+  }:
+
+  let
+    inherit (lib) assertMsg nameValuePair listToAttrs;
+    rawLines = splitString "\n" configText;
+    nonEmptyLines = filter (line: line != "") rawLines;
+    configLines = filter (line: !(hasPrefix "#" line)) nonEmptyLines;
```
Comment (shlevy) [inline lib/strings.nix:1186]: What if a comment line starts with whitespace, is that legal in a kconfig?

File: lib/strings.nix [hunk 2]
```diff
@@ -1149,6 +1150,68 @@ rec {
...
+    parseConfigLine =
+      line:
+      let
+        parsed = parseConfigLineUnquoted line;
+      in
+      parsed // { value = removeQuote parsed.value; };
+    parseConfigLineUnquoted =
+      line:
+      let
+        parsed = match "(CONFIG_[^=]+)=(.*)$" line;
```
Comment (shlevy) [inline lib/strings.nix:1196]: Is this format for CONFIG keys documented anywhere upstream?

File: lib/strings.nix [hunk 3]
```diff
@@ -1149,6 +1150,68 @@ rec {
...
+      in
+      assert assertMsg (parsed != null) "failed to parse kernel config line '${line}'";
+      nameValuePair (elemAt parsed 0) (elemAt parsed 1);
+    removeQuote =
```
Comment (shlevy) [inline lib/strings.nix:1200]: Does kconfig not distinguish between `CONFIG_FOO=m` and `CONFIG_FOO=“m”`?

File: lib/strings.nix [hunk 4]
```diff
@@ -1149,6 +1150,68 @@ rec {
...
+      => {
+        CONFIG_A = "y";
+        CONFIG_B = "m";
+        CONFIG_C = "text";
+      }
+  */
+  parseLinuxConfig = {
+    configText,
+    # skip ill-formed lines
+    relaxed ? false,
```
Comment (shlevy) [inline lib/strings.nix:1179]: Why do we need this option?

File: pkgs/os-specific/linux/kernel/manual-config.nix [hunk 5]
```diff
@@ -8,15 +8,12 @@ let
...
-      no_firstquote="''${val#\"}";
-      echo '  "'"$key"'" = "'"''${no_firstquote%\"}"'";' >> "$out"
-    done < "${configfile}"
-    echo "}" >> $out
-  '').outPath;
+  readConfig = configfile:
+    lib.strings.parseLinuxConfig {
+      # IFD could happen in `builtins.readFile`
+      configText = builtins.readFile configfile;
+      relaxed = true;
```
Comment (shlevy) [inline pkgs/os-specific/linux/kernel/manual-config.nix:15]: Why?

**Review by shlevy — COMMENTED** [review]
Overall idea seems sound if performance is comparable as you say, a few comments though

**Comment by linyinfeng** [comment]
> Hmm not a fan of this.
> 
> Here's an alternative idea: How about exposing the [tool to convert a kernel config to a `config.nix`](https://github.com/NixOS/nixpkgs/blob/9ce47d84cddede3ee9330f63f458e60e267f939c/pkgs/os-specific/linux/kernel/manual-config.nix#L11-L19) somewhere under `pkgs.*`, such that users can commit a `config.nix` instead. And from a `config.nix` it's very straightforward to generate a kernel config again.

@infinisil Thank you! I think it is a good idea. When the config file is available in evaluation time, something like `pkgs.linuxConfig2nix` (maybe another name, I'm not sure) with `buildLinux` is sufficient. We can make a `config.nix` and pass the config expression to the `structuredConfig` option of `buildLinux`, and turn `enableCommonConfig = false`.

<details>
<summary>
We can not use such config.nix with linuxManualConfig, so use buildLinux instead.
</summary>

In `linuxManualConfig`, the `config` argument is never used in the real build process, the function just extends `config` with some utils functions like `isYes` and `isModule` etc, then expose it to `passthru.config`.
This behavior makes `linuxManualConfig` lacking some features.
1. Manually passing `config` is almost useless. It only makes the `config` not consistent with `configfile`.
2. `extraConfig`s in `kernelPatches` are not respected.
</details>

While I still want to claim that a nix parse function for linux config strings is useful, since it eliminates the need to run some utils like `linuxConfig2nix`, I decided to convert this PR back to a draft.

Thank you @shlevy for the detailed review! The major problem is that I did not find documentation for the `.config` file, the behavior of the current implementation still differs from the kernel's build system. So, adding it to `lib.strings` might not be a good idea.

For example, the kernel will threat the following config, as `CONFIG_A="test"`, while my implementation will parse it to `CONFIG_A = "\"test\" blabla";`.
```
CONFIG_A="test" blabla
```

<details>
<summary>Answers to @shlevy's comments</summary>

> What if a comment line starts with whitespace, is that legal in a kconfig?

No, it will emit a warning.
```console
$ make oldconfig
.config:5:warning: unexpected data:   #
```

> Is this format for CONFIG keys documented anywhere upstream?
> Does kconfig not distinguish between CONFIG_FOO=m and CONFIG_FOO=“m”?

I did not find documentation for the `.config` file.

I follow the original bash version, it ignores all lines not begin with `CONFIG_`, and it removes the first layer of quotes.

> Why do we need this option?
(The option is relaxed = true)

I add this option to ignore some ill-formed lines to maintain compatibility with the original bash version.
I found that some kernel config files contain ill-formed lines, for example, <https://github.com/NickCao/nixos-riscv/blob/8c8a79438ec8ee640e59d00aa5878607495a89de/prebuilt/duo-256-kernel-config.txt#L691>.

```
CONFIG_FAILOVER is not set
```
(no `#` in front)

The original bash version actually accepts this, while it wrongly parses it to `{ "CONFIG_FAILOVER is not set" = ""; }`.

</details>

**Comment by linyinfeng** [comment]
Last week I investigated how kernel parses the `.config` file and wrote two simple tools to convert `.config` to structured config accepted by `buildLinux` (<https://github.com/linyinfeng/conf2nix>).

I realized that there are gaps between `.config` and structured config. **Without parsing the full Kconfig source, there is no way to properly parse `.config` and convert it to structured config.** So I'm going to close this PR, and try to maintain some separate tools to help developers maintaining huge structured config.

1. In nixpkgs, structured config is used in [generate-config.pl](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/kernel/generate-config.pl) to interactively answer questions in make config (precisely make oldaskconfig). While the `.config` file is not directly related to prompt and answer, whether a symbol will be written to `.config` is controlled by a complex `SYMBOL_WRITE` flag. Also, the question information is only available in Kconfig source.
2. Parsing process of `.config` is directed by the type of symbols, which only available in the target Kconfig source. <https://github.com/torvalds/linux/blob/20cb38a7af88dc40095da7c2c9094da3873fea23/scripts/kconfig/confdata.c#L236>

**Pull request closed without merging**